### PR TITLE
[#117] Warn on duplicate properties

### PR DESCRIPTION
**Pull request by ghillairet**
See #117

### CODE REVIEW

File: com.reprezen.swagedit/src/com/reprezen/swagedit/validation/Validator.java [hunk 1]
```diff
@@ -56,21 +66,85 @@
...
+			for (NodeTuple tuple: n.getValue()) {
+				Node keyNode = tuple.getKeyNode();
+
+				if (keyNode.getNodeId() == NodeId.scalar) {
+					String key = ((ScalarNode) keyNode).getValue();
+					if (duplicates.containsKey(key)) {
+						duplicates.get(key).add(keyNode);
+					} else {
+						duplicates.put(key, Sets.newHashSet(keyNode));
+					}
```
Comment (tfesenko) [inline com.reprezen.swagedit/src/com/reprezen/swagedit/validation/Validator.java]: I don't know why you don't like Guava, but sometimes it's very useful. A `Multimap` would save us a four lines of the if-else here:

``` java
Multimap<String, Node> duplicates = HashMultimap.<String, Node>create();
<...>
duplicates.put(key, keyNode);
```

File: com.reprezen.swagedit.tests/src/com/reprezen/swagedit/tests/ValidatorTest.xtend [hunk 2]
```diff
@@ -187,4 +189,89 @@ class ValidatorTest {
...
+
+		document.set(content)
+		val errors = validator.validate(document)
+
+		assertEquals(2, errors.size())
+		assertThat(errors, hasItems(
+			new SwaggerError(10, IMarker.SEVERITY_WARNING, String.format(Messages.error_duplicate_keys, "in")),
+			new SwaggerError(11, IMarker.SEVERITY_WARNING, String.format(Messages.error_duplicate_keys, "in"))
+		))
+	}
```
Comment (tfesenko) [inline com.reprezen.swagedit.tests/src/com/reprezen/swagedit/tests/ValidatorTest.xtend:299]: I would also add a test for a case when the value is an object, for example here `responses` defined twice within the same pathItem:

``` yaml
swagger: '2.0'
info:
  version: 0.0.0
  title: Simple API
paths:
  /foo/{bar}:
    get:
      responses:
        '200':
          description: OK
      responses:
        '201':
          description: OK
```

It works as expected - validator shows a warning, but we don't have a test for it.

File: com.reprezen.swagedit/src/com/reprezen/swagedit/validation/Validator.java [hunk 3]
```diff
@@ -56,21 +66,85 @@
...
+			for (String key: duplicates.keySet()) {
+				Set<Node> values = duplicates.get(key);
+				if (values.size() > 1) {
+					for (Node duplicate: values) {
+						errors.add(new SwaggerError(
+								duplicate.getStartMark().getLine() + 1, 
+								IMarker.SEVERITY_WARNING, 
+								String.format(Messages.error_duplicate_keys, key)));
+					}
+				}
```
Comment (tfesenko) [inline com.reprezen.swagedit/src/com/reprezen/swagedit/validation/Validator.java:138]: I am thinking of a burden on performance. We have two passes here:
1. to create a map
2. to check that every key has a single value.
@ghillairet , let's think how we can improve it. Maybe, use a map that would throw an exception if you add a value for the same key twice and catch an exception.

File: com.reprezen.swagedit/src/com/reprezen/swagedit/validation/Validator.java [hunk 4]
```diff
@@ -56,21 +66,85 @@
...
+		Set<SwaggerError> errors = Sets.newHashSet();
+		
+		if (node.getNodeId() == NodeId.mapping) {
+			MappingNode n = (MappingNode) node;
+
+			Map<String, Set<Node>> duplicates = Maps.newHashMap();
+			for (NodeTuple tuple: n.getValue()) {
+				Node keyNode = tuple.getKeyNode();
+
+				if (keyNode.getNodeId() == NodeId.scalar) {
```
Comment (tfesenko) [inline com.reprezen.swagedit/src/com/reprezen/swagedit/validation/Validator.java]: I wonder why we need a check for `scalar` here

File: com.reprezen.swagedit.tests/src/com/reprezen/swagedit/tests/ValidatorTest.xtend (hunk 2 above)
Comment (tfesenko) [inline com.reprezen.swagedit.tests/src/com/reprezen/swagedit/tests/ValidatorTest.xtend:299]: Thanks for the tests!

File: com.reprezen.swagedit/src/com/reprezen/swagedit/validation/Validator.java (hunk 3 above)
Comment (ghillairet) [inline com.reprezen.swagedit/src/com/reprezen/swagedit/validation/Validator.java:138]: If you throw the exception on the first key encounter twice, how do you handle cases where an object has multiple different duplicate keys?
What I can do is collect all duplicates in a single map and resolve duplicates once after the document is parsed.

File: com.reprezen.swagedit/src/com/reprezen/swagedit/validation/Validator.java [hunk 5]
```diff
@@ -101,50 +97,65 @@
...
-								String.format(Messages.error_duplicate_keys, key)));
-					}
+	/*
+	 * This method iterates through the YAML tree to collect the pairs of Node x String 
+	 * representing an object and one of it's keys. Each pair is associated to a Set of 
+	 * Nodes which contains all nodes being a key to the pair's Node and having for value 
+	 * the pair's key.
+	 * Once the iteration is done, the resulting map should be traversed. Each pair having 
+	 * more than one element in its associated Set are duplicate keys. 
+	 */
```
Comment (tfesenko) [inline com.reprezen.swagedit/src/com/reprezen/swagedit/validation/Validator.java:129]: This is a comment, not a Java doc. I wonder if it was your intention.
Similar with other methods in this class.

**Comment by tfesenko** [comment]
Code looks good!
the performance is good too - it took 1-2 milliseconds on a valid 2800 line model to pass `checkDuplicateKey`

**Pull request merged**